pal/src/cruntime/misc.cpp

```diff
@@ -251,57 +251,241 @@ PAL_qsort(void *base, size_t nmemb, size_t size,
+    SORT2(1, 4)  \
+    SORT2(2, 5)  \
+    SORT2(2, 4)  \
+    SORT2(1, 3)  \
+    SORT2(2, 3)
+
+#define SORT7() \
+    SORT2(1, 2) \
+    SORT2(0, 2) \
+    SORT2(0, 1) \
+    SORT2(3, 4) \
+    SORT2(5, 6) \
+    SORT2(3, 5) \
+    SORT2(4, 6) \
+    SORT2(4, 5) \
+    SORT2(0, 4) \
+    SORT2(0, 3) \
+    SORT2(1, 5) \
+    SORT2(2, 6) \
+    SORT2(2, 5) \
+    SORT2(1, 3) \
+    SORT2(2, 4) \
+    SORT2(2, 3)
+
+#define SORT8() \
+    SORT2(0, 1) \
+    SORT2(2, 3) \
+    SORT2(0, 2) \
+    SORT2(1, 3) \
+    SORT2(1, 2) \
+    SORT2(4, 5) \
+    SORT2(6, 7) \
+    SORT2(4, 6) \
+    SORT2(5, 7) \
+    SORT2(5, 6) \
+    SORT2(0, 4) \
+    SORT2(1, 5) \
+    SORT2(1, 4) \
+    SORT2(2, 6) \
+    SORT2(3, 7) \
+    SORT2(3, 6) \
+    SORT2(2, 4) \
+    SORT2(3, 5) \
+    SORT2(3, 4)
+// Sorting Networks END *****
+
+__attribute__((no_instrument_function))
+static void
+cc_qsort_r(void *base_, size_t nmemb, size_t size,
+    int (__cdecl *compar )(void*, const void *, const void *), void *thunk) noexcept
+{
+    char *base = (char *) base_;
+
+    if (nmemb < 9) // use sorting networks for members <= 2^3
+    {
+        switch(nmemb)
+        {
+            case 8:
+                SORT8()
+                break;
+            case 7:
+                SORT7()
+                break;
+            case 6:
+                SORT6()
+                break;
+            case 5:
+                SORT5(0, 1, 2, 3, 4)
+                break;
+            case 4:
+                SORT4()
+                break;
+            case 3:
+                SORT3()
+                break;
+            case 2:
+                SORT2(0, 1)
+                break;
+            default:
+                break; // nothing to sort
+        }
+
+        return;
+    }
+
+    // sort find for median. `compare` is expectedly more expensive than `swap`
+    SORT5(0, (nmemb / 4), (nmemb / 2), ((3 * nmemb) / 4), (nmemb - 1));
+    // guess the median is in the middle now. (from the list above)
+    size_t pos = 0;
+    size_t pivot = (nmemb - 1) * size;
+    // make last element the median(pivot)
+    cc_swap(base + pivot, base + ((nmemb / 2) * size), size);
+    // standard qsort pt. below
+    for (size_t i = 0; i < pivot; i += size)
+    {
+        if (compar(thunk, base + i, base + pivot) <= 0)
+        {
+            cc_swap(base + i, base + (pos * size), size);
+            pos++;
+        }
+    }
+    // issue the last change
+    cc_swap(base + (pos * size), base + pivot, size);
+
+    if (pos >= nmemb - 1) {
+        return; // looks like it was either all sorted OR nothing to sort
+    }
+
+    cc_qsort_r(base, pos++, size, compar, thunk);
+    cc_qsort_r(base + (pos * size), nmemb - pos, size, compar, thunk);
 }
-#endif
 
 PALIMPORT
 void __cdecl
 PAL_qsort_s(void *base, size_t nmemb, size_t size,
-    int (__cdecl *compar )(void*, const void *, const void *), void* context)
+    int (__cdecl *compar )(void*, const void *, const void *), void* context) noexcept
```

Comment:
> std::terminate is called immediately

The `immediately` part is IMHO open to misinterpretation. Otherwise, that would be similarly expensive. Because, there has to be an additional code to track that, right? However AFAIK compiler doesn't put anything there (when noexcept is set) and exception slips to a caller. There, we have the exception handler in place.

Agree this is a trick and saves ~4%